### PR TITLE
[rtl872x] spi_hal / gpio_hal changes for neopixel

### PR DESCRIPTION
**Pull request by technobly**
### Problem

Desirable to add P2 / Photon 2 platform support in the NeoPixel library.

### Solution

- Implement NeoPixel support with SPI/SPI1 MOSI output
- Make a few changes for spi_hal / gpio_hal to allow SCK/SS/MISO pins to be used as GPIO

### References

- https://github.com/technobly/Particle-NeoPixel/pull/45

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [x] Run unit/integration/application tests on device
- [x] N/A - Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: hal/src/rtl872x/spi_hal.cpp [hunk 1]
```diff
@@ -878,18 +892,24 @@ void hal_spi_init(hal_spi_interface_t spi) {
 }
 
 void hal_spi_begin(hal_spi_interface_t spi, uint16_t pin) {
+    // TODO: remove after adding Wiring layer implementation, if required
```
Comment (avtolstoy) [inline hal/src/rtl872x/spi_hal.cpp:895]: Should probably just allow passing `nullptr` for default behavior?

File: hal/inc/hal_dynalib_spi.h [hunk 2]
```diff
@@ -50,7 +50,7 @@ DYNALIB_FN(7, hal_spi, hal_spi_init, void(hal_spi_interface_t))
 DYNALIB_FN(8, hal_spi, hal_spi_is_enabled, bool(hal_spi_interface_t))
 DYNALIB_FN(9, hal_spi, hal_spi_info, void(hal_spi_interface_t, hal_spi_info_t*, void*))
 DYNALIB_FN(10, hal_spi, hal_spi_transfer_dma, void(hal_spi_interface_t, const void*, void*, uint32_t, hal_spi_dma_user_callback))
-DYNALIB_FN(11, hal_spi, hal_spi_begin_ext, void(hal_spi_interface_t, hal_spi_mode_t, uint16_t, void*))
+DYNALIB_FN(11, hal_spi, hal_spi_begin_ext, void(hal_spi_interface_t, hal_spi_mode_t, uint16_t, const hal_spi_config_t*))
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/inc/hal_dynalib_spi.h:53]: Correct, we cannot add new arguments, but like in this case we can change from `void*` to another pointer type, the types are compatible, the sizes are the same and applications built for older Device OS versions will correctly pass `nullptr` here meaning 'use defaults'.

[the remark this answers is not in the document]
Comment (scott-brust) [reply hal/inc/hal_dynalib_spi.h:53]: Ok, and in this case, we have a pointer to a versioned structure, so if we need to pass new parameters we can extend that structure

File: hal/shared/spi_lock.h [hunk 3]
```diff
@@ -50,7 +50,12 @@ class SpiConfigurationLock {
             hal_spi_init(spi_);
             // Make sure the SPI peripheral is initialized with default settings
             hal_spi_acquire(spi_, nullptr);
-            hal_spi_begin_ext(spi_, SPI_MODE_MASTER, PIN_INVALID, nullptr);
+            const hal_spi_config_t spi_config = {
+                .size = sizeof(hal_spi_config_t),
+                .version = HAL_SPI_CONFIG_VERSION_1,
+                .flags = (uint32_t)HAL_SPI_CONFIG_FLAG_NONE
+            };
+            hal_spi_begin_ext(spi_, SPI_MODE_MASTER, PIN_INVALID, &spi_config);
```
[the remark this answers is not in the document]
Comment (technobly) [reply hal/shared/spi_lock.h:58]: I'll switch this back to passing the nullptr... the idea is to continue to support default initialization here.

File: hal/inc/hal_dynalib_spi.h [hunk 4]
```diff
@@ -50,7 +50,7 @@ DYNALIB_FN(7, hal_spi, hal_spi_init, void(hal_spi_interface_t))
 DYNALIB_FN(8, hal_spi, hal_spi_is_enabled, bool(hal_spi_interface_t))
 DYNALIB_FN(9, hal_spi, hal_spi_info, void(hal_spi_interface_t, hal_spi_info_t*, void*))
 DYNALIB_FN(10, hal_spi, hal_spi_transfer_dma, void(hal_spi_interface_t, const void*, void*, uint32_t, hal_spi_dma_user_callback))
-DYNALIB_FN(11, hal_spi, hal_spi_begin_ext, void(hal_spi_interface_t, hal_spi_mode_t, uint16_t, void*))
+DYNALIB_FN(11, hal_spi, hal_spi_begin_ext, void(hal_spi_interface_t, hal_spi_mode_t, uint16_t, hal_spi_config_t*))
```
Comment (avtolstoy) [inline hal/inc/hal_dynalib_spi.h:53]: `const hal_spi_config_t*` to indicate that it's a r/o argument.

File: user/tests/wiring/api/spi.cpp [hunk 5]
```diff
@@ -209,11 +209,15 @@ test(spi_hal_backwards_compatibility)
     SPI_Mode mode = SPI_MODE_MASTER;
     HAL_SPI_TransferStatus status;
     HAL_SPI_AcquireConfig config;
+    hal_spi_config_t spi_config = {};
+    spi_config.size = sizeof(spi_config);
+    spi_config.version = HAL_SPI_CONFIG_VERSION_1;
```
Comment (avtolstoy) [inline user/tests/wiring/api/spi.cpp:214]: Prefer to use latest (`HAL_SPI_CONFIG_VERSION`). It's fine if some fields are left as defaults (zero-filled), that should be correctly handled.

**Review by avtolstoy — APPROVED** [review]

**Pull request merged**